lib/compiler/src/beam_call_types.erl

```diff
@@ -492,78 +512,46 @@ types(erlang, 'node', [_]) ->
-                   #t_tuple{elements=Es0,size=Size}=T} when Index >= 1 ->
-                      %% This is an exact index, update the type of said
-                      %% element or return 'none' if it's known to be out of
-                      %% bounds.
-                      Es = beam_types:set_tuple_element(Index, ArgType, Es0),
-                      case T#t_tuple.exact of
-                          false ->
-                              T#t_tuple{size=max(Index, Size),elements=Es};
-                          true when Index =< Size ->
-                              T#t_tuple{elements=Es};
-                          true ->
-                              none
-                      end;
-                  {#t_integer{elements={Min,Max}},
-                   #t_tuple{elements=Es0,size=Size}=T}
-                    when is_integer(Min), Min >= 1, is_integer(Max) ->
-                      %% We know this will land between Min and Max, so kill
-                      %% the types for those indexes.
-                      Es = discard_tuple_element_info(Min, Max, Es0),
-                      case T#t_tuple.exact of
-                          false ->
-                              T#t_tuple{elements=Es,size=max(Min, Size)};
-                          true when Min =< Size ->
-                              T#t_tuple{elements=Es,size=Size};
-                          true ->
-                              none
-                      end;
-                  {_,#t_tuple{}=T} ->
-                      %% Position unknown, so we have to discard all element
-                      %% information.
+types(erlang, element, [Pos, Tuple0]) ->
+    PosRange = #t_integer{elements={1,?MAX_TUPLE_SIZE}},
+    case meet(Pos, PosRange) of
+        #t_integer{elements={Index,Index}} when Index >= 1 ->
+            case normalize(meet(Tuple0, #t_tuple{size=Index})) of
+                #t_tuple{elements=Es}=Tuple ->
+                    RetType = beam_types:get_tuple_element(Index, Es),
+                    sub_unsafe(RetType, [PosRange, Tuple]);
+                none ->
+                    sub_unsafe(none, [PosRange, #t_tuple{}])
+            end;
+        _ ->
+            sub_unsafe(any, [PosRange, #t_tuple{}])
+    end;
+types(erlang, setelement, [PosType, TupleType0, ArgType]) ->
+    PosRange = #t_integer{elements={1,?MAX_TUPLE_SIZE}},
+    TupleType = meet(TupleType0, #t_tuple{size=1}),
+    RetType = case {meet(PosType, PosRange), normalize(TupleType)} of
+                  {#t_integer{elements={Same,Same}}, #t_tuple{}} ->
+                      beam_types:update_tuple(TupleType, [{Same, ArgType}]);
+                  {#t_integer{}, #t_tuple{}=T} ->
                       T#t_tuple{elements=#{}};
-                  {#t_integer{elements={Min,_Max}},_} ->
-                      #t_tuple{size=Min};
-                  {_,_} ->
-                      #t_tuple{}
+                  {_, _} ->
+                      none
```

Comment:
Not covered by the compiler test suite. This is one of sixteen lines not covered in this module. All lines used to be covered.